src/include/migraphx/op/roialign.hpp

```diff
@@ -0,0 +1,321 @@
+            float w4 = ly * lx;
+
+            // save weights and indeces
+            pos_weight pc;
+            pc.pos1            = y_low * width + x_low;
+            pc.pos2            = y_low * width + x_high;
+            pc.pos3            = y_high * width + x_low;
+            pc.pos4            = y_high * width + x_high;
+            pc.w1              = w1;
+            pc.w2              = w2;
+            pc.w3              = w3;
+            pc.w4              = w4;
+            pos_weights[index] = pc;
+        });
+    }
+
+    struct max_pool
+    {
+        double init() { return std::numeric_limits<double>::lowest(); }
+
+        double op(double x, double y)
+        {
+            double m = std::max(x, y);
+            return (m);
+        }
+
+        double final(double x, std::size_t) { return (x); }
+    };
+
+    struct avg_pool
+    {
+        double init() { return 0.0; }
+
+        double op(double x, double y) { return x + y; }
+
+        double final(double x, std::size_t y) { return (y == 0) ? 0.0 : (x / y); }
+    };
+
+    template <class T, class Op>
+    double calc_pooling(const T* data,
+                        int64_t roi_bin_grid_h,
+                        int64_t roi_bin_grid_w,
+                        const std::vector<pos_weight>& pos_weights,
+                        int64_t& index,
+                        Op op) const
+    {
+        double output_val   = op.init();
+        const int64_t count = roi_bin_grid_h * roi_bin_grid_w;
+        for(int64_t iy = 0; iy < roi_bin_grid_h; iy++)
+        {
+            for(int64_t ix = 0; ix < roi_bin_grid_w; ix++)
+            {
+                const auto& pc = pos_weights[index];
+                output_val     = op.op(output_val, pc.w1 * data[pc.pos1]);
+                output_val     = op.op(output_val, pc.w2 * data[pc.pos2]);
+                output_val     = op.op(output_val, pc.w3 * data[pc.pos3]);
+                output_val     = op.op(output_val, pc.w4 * data[pc.pos4]);
+                index += 1;
+            }
+        }
+        output_val = op.final(output_val, count);
+
+        return output_val;
+    }
+
+    argument compute(const shape& output_shape, std::vector<argument> args) const
+    {
+        argument result{output_shape};
+        const auto& out_lens  = output_shape.lens();
+        int64_t n_rois        = out_lens[0];
+        int64_t channels      = out_lens[1];
+        int64_t pooled_height = out_lens[2];
+        int64_t pooled_width  = out_lens[3];
+        const auto& x_lens    = args.at(0).get_shape().lens();
+        auto height           = x_lens[2];
+        auto width            = x_lens[3];
+        auto roi_s            = args.at(1).get_shape();
+
+        visit_all(result, args.at(0), args.at(1))([&](auto output, auto x, auto roi) {
+            const auto* batch_indices = args.at(2).cast<int64_t>();
+            par_for(n_rois, [&](auto n) {
+                const auto* bottom_data  = x.data();
```

Comment:
Use iterators instead: `auto bottom_data = x.begin();`. You will need to update `calc_pooling` to remove the pointer.

Reply:
fixed